sql/atxs/atxs.go

```diff
@@ -324,37 +325,58 @@ func GetBlobSizes(db sql.Executor, ids [][]byte) (sizes []int, err error) {
 }
 
 // LoadBlob loads ATX as an encoded blob, ready to be sent over the wire.
-func LoadBlob(ctx context.Context, db sql.Executor, id []byte, blob *sql.Blob) error {
+func LoadBlob(ctx context.Context, db sql.Executor, id []byte, blob *sql.Blob) (types.AtxVersion, error) {
 	if sql.IsCached(db) {
-		b, err := getBlob(ctx, db, id)
+		type cachedBlob struct {
+			version types.AtxVersion
+			buf     []byte
+		}
+		cacheKey := sql.QueryCacheKey(CacheKindATXBlob, string(id))
+		cached, err := sql.WithCachedValue(ctx, db, cacheKey, func(context.Context) (*cachedBlob, error) {
+			// We don't use the provided blob in this case to avoid
+			// storing references to the underlying slice (subsequent calls would modify it).
+			var blob sql.Blob
+			v, err := getBlob(ctx, db, id, &blob)
+			if err != nil {
+				return nil, err
+			}
+			return &cachedBlob{version: v, buf: blob.Bytes}, nil
+		})
 		if err != nil {
-			return err
+			return 0, err
 		}
-		blob.Bytes = b
-		return nil
+
+		n := len(cached.buf)
+		blob.Bytes = slices.Grow(blob.Bytes[:0], n)[:n]
+		copy(blob.Bytes, cached.buf)
+		return cached.version, nil
 	}
-	return sql.LoadBlob(db, "select atx from atx_blobs where id = ?1", id, blob)
+
+	return getBlob(ctx, db, id, blob)
 }
 
-func getBlob(ctx context.Context, db sql.Executor, id []byte) (buf []byte, err error) {
-	cacheKey := sql.QueryCacheKey(CacheKindATXBlob, string(id))
-	return sql.WithCachedValue(ctx, db, cacheKey, func(context.Context) ([]byte, error) {
-		if rows, err := db.Exec("select atx from atx_blobs where id = ?1",
-			func(stmt *sql.Statement) {
-				stmt.BindBytes(1, id)
-			}, func(stmt *sql.Statement) bool {
-				if stmt.ColumnLen(0) > 0 {
-					buf = make([]byte, stmt.ColumnLen(0))
-					stmt.ColumnBytes(0, buf)
-				}
-				return true
-			}); err != nil {
-			return nil, fmt.Errorf("get %s: %w", types.BytesToHash(id), err)
-		} else if rows == 0 {
-			return nil, fmt.Errorf("%w: atx %s", sql.ErrNotFound, types.BytesToHash(id))
-		}
-		return buf, nil
-	})
+func getBlob(ctx context.Context, db sql.Executor, id []byte, blob *sql.Blob) (types.AtxVersion, error) {
+	var version types.AtxVersion
+	rows, err := db.Exec("select atx, version from atx_blobs where id = ?1",
+		func(stmt *sql.Statement) {
+			stmt.BindBytes(1, id)
+		}, func(stmt *sql.Statement) bool {
+			n := stmt.ColumnLen(0)
+			blob.Bytes = slices.Grow(blob.Bytes[0:], n)[:n]
+			stmt.ColumnReader(0).Read(blob.Bytes)
```

Comment:
Maybe `Blob.fromColumn` method could be renamed to `Blob.FromColumn` and used here?